R/print.data.table.R

```diff
@@ -34,6 +34,8 @@ print.data.table = function(x, topn=getOption("datatable.print.topn"),
         ( length(SYS) >= 3L && is.symbol(thisSYS <- SYS[[length(SYS)-2L]][[1L]]) &&
           as.character(thisSYS) == 'source') || # suppress printing from source(echo = TRUE) calls, #2369
         ( length(SYS) > 3L && is.symbol(thisSYS <- SYS[[length(SYS)-3L]][[1L]]) &&
+          as.character(thisSYS) %chin% mimicsAutoPrint ) ||
```

Comment:
I don't really understand the issue/fix, so comments would be super helpful here.
Could you please add somethign like `# suppress printing in knitr, #6509 `
and maybe some explanation about where the magic number 6 comes from?

Reply:
Does ffeb415 help?

The original problem is that `[.data.table` would like to return normally for `DT[<any read-only operation>]` but _invisibly_ for `DT[,<var> := <assignment by reference>]`.

Unfortunately, when evaluating calls to primitives, [R checks a special flag for whether to restore the visibility](https://github.com/r-devel/r-svn/blob/734e1434d3b229a008e73e3dfd4a27d733564e0e/src/main/eval.c#L1228-L1242), and the primitive `` `[` `` has [this flag](https://github.com/r-devel/r-svn/blob/734e1434d3b229a008e73e3dfd4a27d733564e0e/src/main/names.c#L64) [set to 0](https://github.com/r-devel/r-svn/blob/734e1434d3b229a008e73e3dfd4a27d733564e0e/src/main/names.c#L107) (force `R_Visible = TRUE`), so there doesn't seem to be a way to return `invisible(anything)` from `[.data.table`.

To achieve something like that, `data.table` remembers the address of the last data.table on which a by-reference operation has been computed. `print.data.table` checks if the address of the argument matches the remembered one and, if it suspects it's being called for auto-printing (call stack very shallow *or* called by a specific `knitr` function), it refuses to print anything. After something in the `knitr` stack got upgraded, the check for "am I being called for auto-printing by `knitr`" no longer works. We could also change the check to "is `knit_print.default` _anywhere_ in the call stack, not just at positions -3 or -6" and risk false positives.

We could try asking R core to change the flag on `.Primitive("[")`, but almost always forcing visibility on for results of `` `[` `` is the right thing and it'll take years before we can enjoy one less workaround in the codebase.

Reply:
Thanks for the really good summary, really this should be in the codebase somewhere.